library/src/amd_detail/hipsolver_sparse.cpp

```diff
@@ -0,0 +1,909 @@
+    void prep_input(rocsparse_index_base indbase,
+                    int                  n,
+                    int                  nnz,
+                    int*                 ptr,
+                    int*                 ind,
+                    double*              val,
+                    float*               src_val)
+    {
+        int count;
+        if(indbase == rocsparse_index_base_one)
+        {
+            for(int i = 0; i <= n; i++)
+                ptr[i] -= 1;
+            count = std::min(nnz, ptr[n]);
+            for(int i = 0; i < count; i++)
+                ind[i] -= 1;
+        }
+        else
+            count = std::min(nnz, ptr[n]);
+
+        if(src_val)
+        {
+            for(int i = 0; i < count; i++)
+                val[i] = (double)src_val[i];
+        }
+    }
+    // Copy float values into double array
+    void prep_input(int n, double* val, float* src_val)
+    {
+        if(src_val)
+        {
+            for(int i = 0; i < n; i++)
+                val[i] = (double)src_val[i];
+        }
+    }
+
+    // Convert base zero indices to base one, and copy double values into float array
+    void prep_output(rocsparse_index_base indbase,
+                     int                  n,
+                     int                  nnz,
+                     int*                 ptr,
+                     int*                 ind,
+                     double*              val,
+                     float*               dest_val)
+    {
+        int count;
+        if(indbase == rocsparse_index_base_one)
+        {
+            count = std::min(nnz, ptr[n]);
+            for(int i = 0; i <= n; i++)
+                ptr[i] += 1;
+            for(int i = 0; i < count; i++)
+                ind[i] += 1;
+        }
+        else
+            count = std::min(nnz, ptr[n]);
+
+        if(dest_val)
+        {
+            for(int i = 0; i < count; i++)
+                dest_val[i] = (float)val[i];
+        }
+    }
+    // Copy double values into float array
+    void prep_output(int n, double* val, float* dest_val)
+    {
+        if(dest_val)
+        {
+            for(int i = 0; i < n; i++)
+                dest_val[i] = (float)val[i];
+        }
+    }
+#endif
+};
+
+hipsolverStatus_t hipsolverSpCreate(hipsolverSpHandle_t* handle)
+try
+{
+#ifdef HAVE_ROCSPARSE
+    if(!handle)
+        return HIPSOLVER_STATUS_NOT_INITIALIZED;
+
+    hipsolverSpHandle* sp = new hipsolverSpHandle;
+    rocblas_status     status;
+
+    if((status = rocblas_create_handle(&sp->handle)) != rocblas_status_success)
+    {
+        delete sp;
+        return rocblas2hip_status(status);
+    }
+
+    if((status = rocsolver_create_rfinfo(&sp->rfinfo, sp->handle)) != rocblas_status_success)
+    {
+        rocblas_destroy_handle(sp->handle);
+        delete sp;
+        return rocblas2hip_status(status);
+    }
+
+    int ret;
+    if((ret = cholmod_start(&sp->c_handle)) != true)
```

Comment:
I'm not 100% certain of how the promotions are done for this, but I would avoid comparing int and bool. Also, the `ret` variable isn't used.